cirq/experiments/qubit_characterizations.py

```diff
@@ -0,0 +1,639 @@
+        ground state probability.
+        """
+        return [(num, prob) for num, prob in zip(self._num_cfds_seq,
+                                                 self._gnd_state_probs)]
+
+    def plot(self) -> None:
+        """Plots the average ground state probability vs the number of
+        Cliffords in the RB study.
+        """
+        fig = pyplot.figure()
+        pyplot.plot(self._num_cfds_seq, self._gnd_state_probs, 'ro-',
+                    figure=fig)
+        pyplot.xlabel(r"Number of Cliffords")
+        pyplot.ylabel('Ground State Probability')
+
+
+class TomographyResult:
+    """Results from a state tomography experiment."""
+
+    def __init__(self, density_matrix: np.ndarray):
+        """
+        Args:
+            density_matrix: The density matrix obtained from tomography.
+        """
+        self._density_matrix = density_matrix
+
+    @property
+    def data(self) -> np.ndarray:
+        """Returns an n^2 by n^2 complex matrix representing the density
+        matrix of the n-qubit system.
+        """
+        return self._density_matrix
+
+    def plot(self) -> None:
+        """Plots the real and imaginary parts of the density matrix as two
+        3D bar plots.
+        """
+        _plot_density_matrix(self._density_matrix)
+
+
+def rabi_oscillations(sampler: sim.SimulatesSamples, qubit: devices.GridQubit,
+                      final_angle: float, number_shots: int,
+                      number_points: int) -> RabiResult:
+    """Runs a Rabi oscillation experiment.
+
+    Rotates a qubit around the x-axis of the Bloch sphere by a sequence of Rabi
+    angles evenly spaced between 0 and final_angle. For each rotation,
+    repeat the circuit number_shots times and measure the average probability
+    of the qubit being in the |1> state.
+
+    Args:
+        sampler: The quantum engine or simulator to run the circuits.
+        qubit: The qubit under test.
+        final_angle: The final Rabi angle in units of pi.
+        number_shots: The number of repetitions of the circuit for each
+            Rabi angle.
+        number_points: The number of Rabi angles.
+
+    Returns:
+        A RabiResult object that stores and plots the result.
+    """
+    theta = value.Symbol('theta')
+    circuit = circuits.Circuit.from_ops(ops.X(qubit) ** theta)
+    circuit.append(ops.measure(qubit, key='z'))
+    sweep = study.Linspace(key='theta', start=0.0, stop=final_angle,
+                           length=number_points)
+    results = sampler.run_sweep(circuit, params=sweep, repetitions=number_shots)
+    half_turns = np.linspace(0.0, final_angle, number_points)
+    excited_state_probs = np.zeros(number_points)
+    for i in range(number_points):
+        excited_state_probs[i] = np.mean(results[i].measurements['z'])
+
+    return RabiResult(half_turns, excited_state_probs)
+
+
+def single_qubit_randomized_benchmarking(sampler: sim.SimulatesSamples,
```

Comment:
If you can pick reasonable defaults, do so.

Force keyword parameters after the qubit parameter.

Use `num_` instead of `number[s]_` to be more consistent with the rest of the library. `repetitions` instead of `number_shots`.

Reply:
Done